group `credential_response_encryption` parameters in the Issuer Metadata (#153)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

The Client MAY request encrypted responses by providing its encryption parameters in the Credential Request.

> The Credential Issuer MAY require encrypted responses by including `credential_response_encryption` object in the Credential Issuer Metadata.

Reviewer: ```suggestion
The Credential Issuer MAY require encrypted responses by including `credential_response_encryption` in the Credential Issuer Metadata.
```

Reviewer: If we need to refer to this, we should use the word "parameter" as in Issuer Metadata parameter, Authorization Server metadata parameter, etc. "metadata values" might be also fine.

Reviewer 2: ```suggestion
The Credential Issuer MAY require encrypted responses by including `credential_response_encryption` parameter in the Credential Issuer Metadata.
```

* `credential_endpoint`: REQUIRED. URL of the Credential Issuer's Credential Endpoint. This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components.
* `batch_credential_endpoint`: OPTIONAL. URL of the Credential Issuer's Batch Credential Endpoint. This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Batch Credential Endpoint.
* `deferred_credential_endpoint`: OPTIONAL. URL of the Credential Issuer's Deferred Credential Endpoint. This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Deferred Credential Endpoint.

> * `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the (Batch) Credential Response. If this request element is not present, the (Batch) Credential Response returned should not be encrypted.

Reviewer 2: I think the current description is not sufficient. please expand/clarify.

```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the (Batch) Credential Response on top of TLS. If this object is present, the Credential Issuer expects (Batch) Credential Response to be encrypted. If this object is omitted, (Batch) Credential Response should not be encrypted.
```

> * `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the (Batch) Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and/or Batch Credential Response to be encrypted. If this object is omitted, Credential  and/or Batch Credential Response should not be encrypted.

Author: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential  and Batch Credential Response should not be encrypted.
```

Reviewer 2: please also add something about "Object containing... that consists of the following parameters"

> * `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer supports encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential and Batch Credential Response should not be encrypted.

Reviewer 3: The word 'expects' in the latter two sentences reads like a requirement and the sentences don't otherwise add meaningful info.

```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer supports encryption of the Credential and Batch Credential Response on top of TLS. 
```

Reviewer 2: agreed to remove this sentence and reincarnate the boolean that the issuer can use to indicate that the encryption is required. will potenitally lead to a sentence or two in this parameter description related to the reincarnated boolean. @paulbastian

Author: applied in latest commit

> * `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential  and Batch Credential Response should not be encrypted.

Author: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential  and Batch Credential Response should not be encrypted. If present, the object consists of exactly two parameters:
```

Reviewer 2: not sure why this is needed..?

Author: I think you requested this in another conversation in this PR actually.

> * `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential  and Batch Credential Response should not be encrypted.

Reviewer: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the Credential and Batch Credential Response on top of TLS. If this parameter is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this parameter is omitted, Credential  and Batch Credential Response should not be encrypted.
```

Reviewer: See above the reasoning.

Author: the phrase is removed in the latest proposal

> * `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer requires encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential  and Batch Credential Response should not be encrypted.

Reviewer 2: trying to address Brian's comment regarding the word "required"

```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer supports encryption of the Credential and Batch Credential Response on top of TLS. If this object is present, the Credential Issuer expects Credential and Batch Credential Response to be encrypted. If this object is omitted, Credential and Batch Credential Response should not be encrypted.
```

  * `alg_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`alg` values) [@!RFC7518] supported by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].

>   * `enc_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`enc` values) [@!RFC7518] supported by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].

Reviewer 4: I prefer the traditional pattern $whatever`_values_supported` since it was consolidated though years in OAuth 2.0 and OpenID Connect specs

Author: I don't understand. Are you voting to reject the PR?

>   * `enc_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`enc` values) [@!RFC7518] supported by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].

Reviewer: We should probably add something on supported curves. Something like the following ...

```suggestion
  * `enc_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`enc` values) [@!RFC7518] supported by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].
  * `jwk_values_supported`: REQUIRED. Array containing a list of objects describing the supported JWK values by the Credential and/or Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].  Each object MUST contain a (`kty` value) [@!RFC7517] and an OPTIONAL (`crv` value) [@!RFC8037]. For example, to indicate the issuer supports ECDH-ES with curve P-256, `jwk_values_supported` includes the following object: `{ "kty":"EC", "crv":"P-256" }`.
```

Reviewer: Should we add this info about public key types and curves. Otherwise, wallets won't know which public key type they have to provide.

Reviewer 3: This is an interesting question/point, which makes me wonder about other similar places where this kind of info might be useful or needed.

Reviewer 2: I agree it is an interesting point. might be worth opening an issue and addressing this in a separate PR...

Author: See my point above, is this solvable by Fully Specified Algorithms?

Reviewer: IANA does not have fully specified JWE algorithms like that. I suggest we create an issue and move my language to a separate PR.

Reviewer: Created an issue here #168

* `notification_endpoint`: OPTIONAL. URL of the Credential Issuer's Notification Endpoint as defined in (#notification_endpoint). This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not support the Notification Endpoint.
* `credential_response_encryption`: OPTIONAL. Object containing information whether the Credential Issuer supports encryption of the Credential and Batch Credential Response on top of TLS.
  * `alg_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`alg` values) [@!RFC7518] supported by the Credential and Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].
  * `enc_values_supported`: REQUIRED. Array containing a list of the JWE [@!RFC7516] encryption algorithms (`enc` values) [@!RFC7518] supported by the Credential and Batch Credential Endpoint to encode the Credential or Batch Credential Response in a JWT [@!RFC7519].

>   * `encryption_required`: REQUIRED. Boolean value specifying whether the Credential Issuer requires the additional encryption on top of TLS for the Credential Response. If the value is `true`, the Credential Issuer requires the encryption for every Credential Response and therefore the Wallet MUST provide encryption keys in the Credential Request. If the value is `false`, the Wallet MAY chose whether it provides encryption keys or not.

Reviewer 2: I think it logically makes more sense to bring this parameter before `enc_values_supported` and `alg_values_supported` when describing?
